ML CI: Linux aarch64

adamjstewart opened the pull request:
Reboot of #34300

Hoping there will be fewer build issues this time around.

adamjstewart committed:
ML CI: Linux aarch64

adamjstewart committed:
Add config files

adamjstewart committed:
No aarch64 tag

adamjstewart commented:
Need help with this latest error message

adamjstewart committed:
Don't specify image

adamjstewart commented:
@alalazo @becker33 what does this mean and how do I solve it?

> cc1: error: invalid feature modifier in '-march=armv8.2-a+dotprod'

adamjstewart commented:
Analysis of build errors:

### JAX

```
ERROR: /tmp/spackaoztu89p/4cc548a6d16d1f1b062bf8c46c91a45f/external/org_tensorflow/tensorflow/tsl/python/lib/core/BUILD:78:11: Compiling tensorflow/tsl/python/lib/core/float8.cc failed: (Exit 1): gcc failed: error executing command
...
In file included from /usr/lib/gcc/aarch64-redhat-linux/7/../../../../include/c++/7/math.h:36:0,
                 from bazel-out/aarch64-opt/bin/external/local_config_python/python_include/pyport.h:218,
                 from bazel-out/aarch64-opt/bin/external/local_config_python/python_include/Python.h:38,
                 from external/org_tensorflow/tensorflow/tsl/python/lib/core/numpy.h:35,
                 from external/org_tensorflow/tensorflow/tsl/python/lib/core/float8.cc:17:
/usr/lib/gcc/aarch64-redhat-linux/7/../../../../include/c++/7/cmath: In static member function 'static void tsl::custom_float_internal::UnaryUFunc<InType, OutType, Functor>::Call(char**, const npy_intp*, const npy_intp*, void*) [with InType = tsl::float8_internal::float8_e5m2; OutType = tsl::float8_internal::float8_e5m2; Functor = tsl::custom_float_internal::ufuncs::Spacing<tsl::float8_internal::float8_e5m2>]':
/usr/lib/gcc/aarch64-redhat-linux/7/../../../../include/c++/7/cmath:1302:40: internal compiler error: Segmentation fault
   { return __builtin_copysignf(__x, __y); }
                                        ^
```
No idea.

### LLVM

```
/tmp/root/spack-stage/spack-stage-llvm-14.0.6-i75nbdjlmso62n73n6di36ksvguvidb3/spack-src/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64.cpp:376:13: error: 'sve_vl_valid' was not declared in this scope
         if (sve_vl_valid(vg_value * 8)) {
             ^~~~~~~~~~~~
```
No idea. Maybe LLVM maintainers can help? @haampie @skosukhin @trws 

### PyTorch

```
  cc1: error: invalid feature modifier in '-march=armv8.2-a+dotprod'
```
According to https://github.com/google/XNNPACK/issues/1551, we need to either disable XNNPACK or use GCC 10+. I think I can get this working, but may need help upgrading GCC.

### MXNet (CUDA-only)

```
.../linux-ubuntu18.04-aarch64/gcc-7.5.0/cuda-11.8.0-lszmzbgqdjhxte4dq62og3ms7e24xd5m/lib64/libcurand.so: undefined reference to `logf@GLIBC_2.27'
```
No idea.

### LLVM AMD GPU (ROCm-only)

```
CMake Error at /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.6.1-lahtgnfsggfqss7kwskd2h7s5o4gehcj/spack-src/compiler-rt/cmake/Modules/CompilerRTUtils.cmake:367 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.6.1-lahtgnfsggfqss7kwskd2h7s5o4gehcj/spack-src/compiler-rt/CMakeLists.txt:129 (construct_compiler_rt_default_triple)
...
CMake Error at /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.6.1-lahtgnfsggfqss7kwskd2h7s5o4gehcj/spack-src/compiler-rt/cmake/Modules/AddCompilerRT.cmake:766 (get_compiler_rt_output_dir):
  get_compiler_rt_output_dir Function invoked with incorrect arguments for
  function named: get_compiler_rt_output_dir
Call Stack (most recent call first):
  /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.6.1-lahtgnfsggfqss7kwskd2h7s5o4gehcj/spack-src/compiler-rt/unittests/CMakeLists.txt:4 (configure_compiler_rt_lit_site_cfg)
```
No idea. Maintainers are @haampie @renjithravindrankannath @srekolam 

### ROCm SMI Lib (ROCm-only)

```
g++: error: unrecognized command line option '-m64'
g++: error: unrecognized command line option '-msse'; did you mean '-fdse'?
g++: error: unrecognized command line option '-msse2'
```
No idea. Maintainers are @renjithravindrankannath @srekolam

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -767,8 +767,74 @@ ml-linux-x86_64-rocm-build:
+# Machine Learning - Linux aarch64 (CPU)
 ########################################
+.ml-linux-aarch64-cpu:
+  extends: [ ".linux_aarch64" ]
```
Should these also specify the amazonlinux image?

adamjstewart replied:
What's the correct syntax for that? I just copied the template from other ML pipelines, I have absolutely no idea what any of it means.

P.S. Feel free to push to my branch if you want, or suggest edits on this PR.

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -767,8 +767,74 @@ ml-linux-x86_64-rocm-build:
+    SPACK_CI_STACK_NAME: ml-linux-aarch64-cuda
+
+ml-linux-aarch64-cuda-generate:
+  extends: [ ".generate-aarch64", .ml-linux-aarch64-cuda ]
```
```suggestion
  extends: [ ".generate-aarch64", .ml-linux-aarch64-cuda ]
  image: ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09
```

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -767,8 +767,74 @@ ml-linux-x86_64-rocm-build:
+    SPACK_CI_STACK_NAME: ml-linux-aarch64-rocm
+
+ml-linux-aarch64-rocm-generate:
+  extends: [ ".generate-aarch64", .ml-linux-aarch64-rocm ]
```
```suggestion
  extends: [ ".generate-aarch64", .ml-linux-aarch64-rocm ]
  image: ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09
```

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -767,8 +767,74 @@ ml-linux-x86_64-rocm-build:
+    SPACK_CI_STACK_NAME: ml-linux-aarch64-cpu
+
+ml-linux-aarch64-cpu-generate:
+  extends: [ ".generate-aarch64", .ml-linux-aarch64-cpu ]
```
```suggestion
  extends: [ ".generate-aarch64", .ml-linux-aarch64-cpu ]
  image: ghcr.io/spack/e4s-amazonlinux-2:v2023-03-09
```

adamjstewart committed:
Use amazonlinux image

Co-authored-by: kwryankrattiger <80296582+kwryankrattiger@users.noreply.github.com>

adamjstewart commented:
@kwryankrattiger that doesn't seem to have made anything worse, although it doesn't fix any of the above errors.

adamjstewart committed:
Update and require

adamjstewart committed:
GCC is too old

adamjstewart committed:
Fix some builds

adamjstewart commented on var/spack/repos/builtin/packages/rust/package.py:
```diff
@@ -79,6 +79,9 @@ class Rust(Package):
     depends_on("rust-bootstrap@1.74:1.75", type="build", when="@1.75")
 
+    # src/llvm-project/llvm/cmake/modules/CheckCompilerVersion.cmake
+    conflicts("%gcc@:7.3", when="@1.73:", msg="Host GCC version must be at least 7.4")
```
@alecbcs wondering if it's possible to depend on external LLVM installations instead of the vendored copy. Then we don't have to duplicate all compiler version support.

adamjstewart commented on var/spack/repos/builtin/packages/py-onnx/package.py:
```diff
@@ -39,7 +39,6 @@ class PyOnnx(PythonPackage):
 
     # requirements.txt
     depends_on("py-setuptools@64:", type="build")
```
@alalazo the concretizer is selecting setuptools 63 when building py-onnx, which shouldn't be possible: https://gitlab.spack.io/spack/spack/-/jobs/10677665

adamjstewart committed:
xgboost doesn't support old GCC + cuda

adamjstewart commented:
Requires https://github.com/spack/gitlab-runners/pull/49 to build a new image with a modern GCC version.

adamjstewart committed:
Run on newer Ubuntu

adamjstewart commented:
Remaining aarch64 issues are mostly known issues:

* PyTorch: https://github.com/pytorch/pytorch/issues/122169
* TensorFlow: https://github.com/tensorflow/tensorflow/issues/62490

Can anyone look at the llvm-amdgpu build issue on the ROCm stack? @haampie  @renjithravindrankannath  @srekolam

adamjstewart committed:
Remove mxnet

adamjstewart committed:
Try aarch64 range

adamjstewart commented on var/spack/repos/builtin/packages/py-torch/package.py:
```diff
@@ -214,6 +214,8 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("cuda@9.2:11.4", when="@1.6:1.9+cuda", type=("build", "link", "run"))
     depends_on("cuda@9:11.4", when="@:1.5+cuda", type=("build", "link", "run"))
+    # https://github.com/pytorch/pytorch/issues/122169
+    depends_on("cuda@:12.3", when="platform=aarch64:", type=("build", "link", "run"))
```
@alalazo this constraint is also being ignored, the concretizer is choosing cuda 12.4 and it fails.

adamjstewart commented:
TODO: experiment with https://github.com/tensorflow/tensorflow/issues/62490#issuecomment-2077646521 once PyTorch 2.3 is merged.

adamjstewart committed:
Use main branch

adamjstewart committed:
Conflict applies to all targets

adamjstewart committed:
cuda only required when +cuda

adamjstewart commented:
@spackbot run pipeline

adamjstewart committed:
Use tagged version

adamjstewart committed:
Comment out tf-estimator

adamjstewart commented:
@aweits there are only 2 remaining build issues:

* JAX: seems like the same issue with absl that you fixed in https://github.com/tensorflow/tensorflow/issues/62490#issuecomment-2077646521, similar patch may fix this
* TF: a new issues with absl, unclear if it's caused by your patch or a new issue

I know you're busy, but if you find any time to work on this, it would be good to finally get this CI stack merged!

aweits commented:
pending: https://github.com/spack/gitlab-runners/pull/53 and #45437

adamjstewart commented:
TODO: add ROCm

adamjstewart committed:
Merge branch 'develop' into ci/ml-linux-aarch64

adamjstewart committed:
Add ROCm, use newer Ubuntu

adamjstewart commented:
New failures:

### py-tensorflow 2.17.0

Hopefully fixed by #45437, just need to rebase.

### py-jaxlib 0.4.31

```
/usr/lib/gcc/aarch64-linux-gnu/13/include/arm_neon.h(46): error: identifier "__Poly8x8_t" is undefined
  typedef __Poly8x8_t poly8x8_t;
          ^
/usr/lib/gcc/aarch64-linux-gnu/13/include/arm_neon.h(380): error: identifier "__builtin_aarch64_saddlv8qi" is undefined 
    return __builtin_aarch64_saddlv8qi (__a, __b);
           ^
```

https://github.com/jax-ml/jax/issues/22155. Fixed upstream in absl, vendored by XLA, vendored by JAX. May be fixed in newer versions. Worth trying to backport.

### nccl 2.22.3-1

```
/usr/include/aarch64-linux-gnu/bits/math-vector.h(96): error: identifier "__Float32x4_t" is undefined
  typedef __Float32x4_t __f32x4_t;
          ^
```
Haven't yet found any upstream issues.

### gloo 2023-12-03

```
/usr/include/aarch64-linux-gnu/bits/math-vector.h(96): error: identifier "__Float32x4_t" is undefined
  typedef __Float32x4_t __f32x4_t;
          ^
```
Same as above.

### magma 2.8.0

```
  /usr/include/aarch64-linux-gnu/bits/math-vector.h(96): error: identifier
  "__Float32x4_t" is undefined

    typedef __Float32x4_t __f32x4_t;
            ^
```
Same as above. @G-Ragghianti  @luszczek  @stomov any ideas?


### llvm-amdgpu 6.1.2

```
CMake Error at cmake/modules/LLVMExternalProjectUtils.cmake:86 (is_msvc_triple):
  is_msvc_triple Function invoked with incorrect arguments for function
  named: is_msvc_triple
Call Stack (most recent call first):
  runtimes/CMakeLists.txt:81 (llvm_ExternalProject_Add)
  runtimes/CMakeLists.txt:135 (builtin_default_target)


CMake Error at cmake/modules/LLVMExternalProjectUtils.cmake:86 (is_msvc_triple):
  is_msvc_triple Function invoked with incorrect arguments for function
  named: is_msvc_triple
Call Stack (most recent call first): 
  runtimes/CMakeLists.txt:230 (llvm_ExternalProject_Add)
  runtimes/CMakeLists.txt:396 (runtime_default_target)


CMake Error at runtimes/CMakeLists.txt:528 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "check-runtimes".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
```
https://github.com/llvm/llvm-project/issues/78530, fixed by https://github.com/llvm/llvm-project/pull/80071. May need to backport. @haampie  @renjithravindrankannath  @srekolam

luszczek commented:
We don't see this issue on x86 with magma. Maybe @abdelfattah83 seen it with some Grace Hopper builds.

adamjstewart commented:
Correct, this only occurs on aarch64.

aweits commented:
cuda@12.4.1 / cudafe can't handle those types. Issue goes away using cuda@12.5.1 - need to do some more homework here, but that's the quick answer....

aweits commented:
for posterity if anyone searches this and finds this issue -- these kinds of errors happen when gcc's preprocessor emits something that only a version of gcc/g++ can understand - and the next step in the compilation process is a version of cudafe/cudafe++ that doesn't understand it.

```
Apptainer> cat /tmp/test.cu
#include <math.h>

int main()

{}
```

```
Apptainer> nvcc --verbose /tmp/test.cu
#$ _NVVM_BRANCH_=nvvm
#$ _NVVM_BRANCH_SUFFIX_=
#$ _SPACE_= 
#$ _CUDART_=cudart
#$ _HERE_=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin
#$ _THERE_=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin
#$ _TARGET_SIZE_=
#$ _TARGET_DIR_=
#$ _TARGET_DIR_=targets/sbsa-linux
#$ TOP=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/..
#$ NVVMIR_LIBRARY_DIR=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../nvvm/libdevice
#$ LD_LIBRARY_PATH=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../lib:/.singularity.d/libs
#$ PATH=/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../nvvm/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/libxml2-2.10.3-pbsah4muzw2hzvgqtuqy2xccuguk26g6/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/xz-5.4.6-syfksspf4ocxvnxeom6dmyayu76fuksr/bin:/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/libiconv-1.17-aopbjjaoq7nrj4rnmubhewwdt4vmq6n2/bin:/dev/shm/aweits/spack/bin:/opt/venv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
#$ INCLUDES="-I/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../targets/sbsa-linux/include"  
#$ LIBRARIES=  "-L/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../targets/sbsa-linux/lib/stubs" "-L/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../targets/sbsa-linux/lib"
#$ CUDAFE_FLAGS=
#$ PTXAS_FLAGS=
#$ gcc -D__CUDA_ARCH_LIST__=520 -D__NV_LEGACY_LAUNCH -E -x c++ -D__CUDACC__ -D__NVCC__  "-I/dev/shm/aweits/spack/opt/spack/linux-ubuntu24.04-neoverse_v2/gcc-13.2.0/cuda-12.4.1-utxbcpnvw4l5kdpfsaar7rqsk2hulfdu/bin/../targets/sbsa-linux/include"    -D__CUDACC_VER_MAJOR__=12 -D__CUDACC_VER_MINOR__=4 -D__CUDACC_VER_BUILD__=131 -D__CUDA_API_VER_MAJOR__=12 -D__CUDA_API_VER_MINOR__=4 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -include "cuda_runtime.h" "/tmp/test.cu" -o "/scratch.local/tmpxft_003d5a4b_00000000-5_test.cpp4.ii" 
#$ cudafe++ --c++17 --gnu_version=130200 --display_error_number --orig_src_file_name "/tmp/test.cu" --orig_src_path_name "/tmp/test.cu" --allow_managed --unsigned_chars --unsigned_wchar_t  --m64 --parse_templates --gen_c_file_name "/scratch.local/tmpxft_003d5a4b_00000000-6_test.cudafe1.cpp" --stub_file_name "tmpxft_003d5a4b_00000000-6_test.cudafe1.stub.c" --gen_module_id_file --module_id_file_name "/scratch.local/tmpxft_003d5a4b_00000000-4_test.module_id" "/scratch.local/tmpxft_003d5a4b_00000000-5_test.cpp4.ii" 
/usr/include/aarch64-linux-gnu/bits/math-vector.h(96): error: identifier "__Float32x4_t" is undefined
  typedef __Float32x4_t __f32x4_t;
          ^

/usr/include/aarch64-linux-gnu/bits/math-vector.h(97): error: identifier "__Float64x2_t" is undefined
  typedef __Float64x2_t __f64x2_t;
          ^

/usr/include/aarch64-linux-gnu/bits/math-vector.h(106): error: identifier "__SVFloat32_t" is undefined
  typedef __SVFloat32_t __sv_f32_t;
          ^

/usr/include/aarch64-linux-gnu/bits/math-vector.h(107): error: identifier "__SVFloat64_t" is undefined
  typedef __SVFloat64_t __sv_f64_t;
          ^

/usr/include/aarch64-linux-gnu/bits/math-vector.h(108): error: identifier "__SVBool_t" is undefined
  typedef __SVBool_t __sv_bool_t;
          ^

5 errors detected in the compilation of "/tmp/test.cu".
# --error 0x2 --
```

adamjstewart committed:
Merge branch 'develop' into ci/ml-linux-aarch64

adamjstewart commented:
Still need to fix llvm-amdgpu

adamjstewart commented:
py-torchaudio fixed in #46746

aweits commented:
> Still need to fix llvm-amdgpu @haampie @renjithravindrankannath @srekolam. I don't see that file in the ROCm fork so I'm not sure how/where to patch it.

I must be missing something, isn't this just:
```
    patch(
        "https://github.com/ROCm/llvm-project/commit/c651b2b0d9d1393fb5191ac3acfe96e5ecc94bbc.patch?full_index=1",
        sha256="eaf700a5b51d53324a93e5c951bc08b6311ce2053c44c1edfff5119f472d8080",
        when="@6.1.2",
    )
```

adamjstewart commented:
Nope, I'm the one who was missing something. I was looking in `cmake/Modules` instead of `llvm/cmake/modules` 😅 

@aweits want to submit a PR? If not I can get to it later tonight or tomorrow.

adamjstewart commented:
@spackbot run pipeline

adamjstewart commented:
@aweits looks like llvm-amdgpu has a new build error now

adamjstewart commented:
Todd thinks we should just give up on aarch64 + ROCm, so the JAX build is higher priority.

tgamblin commented:
> Todd thinks we should just give up on aarch64 + ROCm, so the JAX build is higher priority.

I'd be thrilled to see it but I don't think it exists!

adamjstewart commented:
We could comment out JAX for now. Once https://github.com/spack/spack/pull/46346 is complete the issue will likely resolve itself. But that PR is pretty far from working.

aweits commented:
i should be able to get some time to poke at this again today...

aweits commented:
here's the tl;dr on a fix for arm/llvm-amdgpu - will put appropriate conditionals on this and submit a pr:
```diff
[aweits@gg-00 spack]$ git diff
diff --git a/var/spack/repos/builtin/packages/llvm-amdgpu/package.py b/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
index 3e39a4445ba..c342098588f 100644
--- a/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
+++ b/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
@@ -222,7 +222,7 @@ def cmake_args(self):
             self.define("LIBCXXABI_ENABLE_STATIC", "ON"),
             self.define("LIBCXXABI_INSTALL_STATIC_LIBRARY", "OFF"),
             self.define("LLVM_ENABLE_RTTI", "ON"),
-            self.define("LLVM_TARGETS_TO_BUILD", "AMDGPU;X86"),
+            self.define("LLVM_TARGETS_TO_BUILD", "AMDGPU;AArch64"),
             self.define("LLVM_AMDGPU_ALLOW_NPI_TARGETS", "ON"),
             self.define("PACKAGE_VENDOR", "AMD"),
             self.define("CLANG_ENABLE_AMDCLANG", "ON"),
```

aweits commented:
got jax building - again, will open pr with appropriate conditionals
```diff
diff --git a/var/spack/repos/builtin/packages/py-jaxlib/jaxxlatsl.patch b/var/spack/repos/builtin/packages/py-jaxlib/jaxxlatsl.patch
new file mode 100644
index 00000000000..e96cc32e263
--- /dev/null
+++ b/var/spack/repos/builtin/packages/py-jaxlib/jaxxlatsl.patch
@@ -0,0 +1,100 @@
+From 8fce7378ed8ce994107568449806cd99274ab22b Mon Sep 17 00:00:00 2001
+From: Andrew Elble <aweits@rit.edu>
+Date: Mon, 21 Oct 2024 19:42:31 -0400
+Subject: [PATCH] patchit
+
+---
+ ...ch-for-Abseil-to-fix-build-on-Jetson.patch | 68 +++++++++++++++++++
+ third_party/xla/workspace.bzl                 |  1 +
+ 2 files changed, 69 insertions(+)
+ create mode 100644 third_party/xla/0001-Add-patch-for-Abseil-to-fix-build-on-Jetson.patch
+
+diff --git a/third_party/xla/0001-Add-patch-for-Abseil-to-fix-build-on-Jetson.patch b/third_party/xla/0001-Add-patch-for-Abseil-to-fix-build-on-Jetson.patch
+new file mode 100644
+index 000000000000..5138a045082b
+--- /dev/null
++++ b/third_party/xla/0001-Add-patch-for-Abseil-to-fix-build-on-Jetson.patch
+@@ -0,0 +1,68 @@
++From 40da87a0476436ca1da2eafe08935787a05e9a61 Mon Sep 17 00:00:00 2001
++From: David Dunleavy <ddunleavy@google.com>
++Date: Mon, 5 Aug 2024 11:42:53 -0700
++Subject: [PATCH] Add patch for Abseil to fix build on Jetson
++
++Patches in https://github.com/abseil/abseil-cpp/commit/372124e6af36a540e74a2ec31d79d7297a831f98
++
++PiperOrigin-RevId: 659627531
++---
++ .../tsl/third_party/absl/nvidia_jetson.patch  | 35 +++++++++++++++++++
++ .../tsl/third_party/absl/workspace.bzl        |  1 +
++ 2 files changed, 36 insertions(+)
++ create mode 100644 third_party/tsl/third_party/absl/nvidia_jetson.patch
++
++diff --git a/third_party/tsl/third_party/absl/nvidia_jetson.patch b/third_party/tsl/third_party/absl/nvidia_jetson.patch
++new file mode 100644
++index 000000000000..5328c3a0d605
++--- /dev/null
+++++ b/third_party/tsl/third_party/absl/nvidia_jetson.patch
++@@ -0,0 +1,35 @@
+++From 372124e6af36a540e74a2ec31d79d7297a831f98 Mon Sep 17 00:00:00 2001
+++From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Bastien?= <frederic.bastien@gmail.com>
+++Date: Thu, 1 Aug 2024 12:38:52 -0700
+++Subject: [PATCH] PR #1732: Fix build on NVIDIA Jetson board. Fix #1665
+++
+++Imported from GitHub PR https://github.com/abseil/abseil-cpp/pull/1732
+++
+++Fix build on NVIDIA Jetson board. Fix #1665
+++
+++This patch is already used by the spark project.
+++I'm fixing this as this break the build of Tensorflow and JAX on Jetson board.
+++Merge 7db2d2ab9fbed1f0fabad10a6ec73533ba71bfff into 6b8ebb35c0414ef5a2b6fd4a0f59057e41beaff9
+++
+++Merging this change closes #1732
+++
+++COPYBARA_INTEGRATE_REVIEW=https://github.com/abseil/abseil-cpp/pull/1732 from nouiz:fix_neon_on_jetson 7db2d2ab9fbed1f0fabad10a6ec73533ba71bfff
+++PiperOrigin-RevId: 658501520
+++Change-Id: If502ede4efc8c877fb3fed227eca6dc7622dd181
+++---
+++ absl/base/config.h | 2 +-
+++ 1 file changed, 1 insertion(+), 1 deletion(-)
+++
+++diff --git a/absl/base/config.h b/absl/base/config.h
+++index 97c9a22a109..ab1e9860a91 100644
+++--- a/absl/base/config.h
++++++ b/absl/base/config.h
+++@@ -926,7 +926,7 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
+++ // https://llvm.org/docs/CompileCudaWithLLVM.html#detecting-clang-vs-nvcc-from-code
+++ #ifdef ABSL_INTERNAL_HAVE_ARM_NEON
+++ #error ABSL_INTERNAL_HAVE_ARM_NEON cannot be directly set
+++-#elif defined(__ARM_NEON) && !defined(__CUDA_ARCH__)
++++#elif defined(__ARM_NEON) && !(defined(__NVCC__) && defined(__CUDACC__))
+++ #define ABSL_INTERNAL_HAVE_ARM_NEON 1
+++ #endif
+++ 
++diff --git a/third_party/tsl/third_party/absl/workspace.bzl b/third_party/tsl/third_party/absl/workspace.bzl
++index 06f75166ce4b..9565a82c3319 100644
++--- a/third_party/tsl/third_party/absl/workspace.bzl
+++++ b/third_party/tsl/third_party/absl/workspace.bzl
++@@ -44,4 +44,5 @@ def repo():
++         system_link_files = SYS_LINKS,
++         strip_prefix = "abseil-cpp-{commit}".format(commit = ABSL_COMMIT),
++         urls = tf_mirror_urls("https://github.com/abseil/abseil-cpp/archive/{commit}.tar.gz".format(commit = ABSL_COMMIT)),
+++        patch_file = ["//third_party/absl:nvidia_jetson.patch"],
++     )
++-- 
++2.31.1
++
+diff --git a/third_party/xla/workspace.bzl b/third_party/xla/workspace.bzl
+index af52e7671507..70481bc970a5 100644
+--- a/third_party/xla/workspace.bzl
++++ b/third_party/xla/workspace.bzl
+@@ -29,6 +29,7 @@ def repo():
+         name = "xla",
+         sha256 = XLA_SHA256,
+         strip_prefix = "xla-{commit}".format(commit = XLA_COMMIT),
++      patch_file = ["//third_party/xla:0001-Add-patch-for-Abseil-to-fix-build-on-Jetson.patch"],
+         urls = tf_mirror_urls("https://github.com/openxla/xla/archive/{commit}.tar.gz".format(commit = XLA_COMMIT)),
+     )
+ 
+-- 
+2.31.1
+
diff --git a/var/spack/repos/builtin/packages/py-jaxlib/package.py b/var/spack/repos/builtin/packages/py-jaxlib/package.py
index 09eb522c56a..ce7f217e8fe 100644
--- a/var/spack/repos/builtin/packages/py-jaxlib/package.py
+++ b/var/spack/repos/builtin/packages/py-jaxlib/package.py
@@ -127,7 +127,7 @@ class PyJaxlib(PythonPackage, CudaPackage, ROCmPackage):
         sha256="4dfb9f32d4eeb0a0fb3a6f4124c4170e3fe49511f1b768cd634c78d489962275",
         when="@:0.4.25",
     )
-
+    patch("jaxxlatsl.patch")
     conflicts(
         "cuda_arch=none",
         when="+cuda",
```

adamjstewart commented:
You can assign me as a reviewer on these and I will merge them

adamjstewart commented:
@afzpatel helped add our x86_64 ROCm pipeline and may be willing to work on an aarch64 ROCm pipeline in a separate PR. I also want to get to ppc64le soon.

aweits commented:
> @afzpatel helped add our x86_64 ROCm pipeline and may be willing to work on an aarch64 ROCm pipeline in a separate PR. I also want to get to ppc64le soon.

I'm certainly willing to help with this.

adamjstewart commented:
rocminfo is next. At this point, we basically need to hack every single package in the ROCm stack. I also want to make sure we're contributing these changes upstream. Let me remove all of the ROCm stuff from this PR and we can tackle ROCm another day.

adamjstewart committed:
Remove ROCm

adamjstewart committed:
Merge branch 'develop' into ci/ml-linux-aarch64

afzpatel commented:
> @afzpatel helped add our x86_64 ROCm pipeline and may be willing to work on an aarch64 ROCm pipeline in a separate PR. I also want to get to ppc64le soon.

Hi @adamjstewart, ARM architecture based build is not supported in ROCm as of today. If there is any change in plan it will be announced through ROCm release notes.
